README.md
## uxcore-list

React list with built-in actionBar, Search & Pagination

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test Coverage][coveralls-image]][coveralls-url]
[![Dependency Status][dep-image]][dep-url]
[![devDependency Status][devdep-image]][devdep-url] 
[![NPM downloads][downloads-image]][npm-url]

[![Sauce Test Status][sauce-image]][sauce-url]

[npm-image]: http://img.shields.io/npm/v/uxcore-list.svg?style=flat-square
[npm-url]: http://npmjs.org/package/uxcore-list
[travis-image]: https://img.shields.io/travis/uxcore/uxcore-list.svg?style=flat-square
[travis-url]: https://travis-ci.org/uxcore/uxcore-list
[coveralls-image]: https://img.shields.io/coveralls/uxcore/uxcore-list.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/uxcore/uxcore-list?branch=master
[dep-image]: http://img.shields.io/david/uxcore/uxcore-list.svg?style=flat-square
[dep-url]: https://david-dm.org/uxcore/uxcore-list
[devdep-image]: http://img.shields.io/david/dev/uxcore/uxcore-list.svg?style=flat-square
[devdep-url]: https://david-dm.org/uxcore/uxcore-list#info=devDependencies
[downloads-image]: https://img.shields.io/npm/dm/uxcore-list.svg
[sauce-image]: https://saucelabs.com/browser-matrix/uxcore-list.svg
[sauce-url]: https://saucelabs.com/u/uxcore-list

![](https://gw.alicdn.com/tps/TB14_QlKVXXXXaTXFXXXXXXXXXX-1012-647.png)


### Development

```sh
git clone https://github.com/uxcore/uxcore-list
cd uxcore-list
npm install
npm run server
```

if you'd like to save your install time，you can use uxcore-tools globally.

```sh
npm install uxcore-tools -g
git clone https://github.com/uxcore/uxcore-list
cd uxcore-list
npm run dep
npm run start
```

### Test Case

```sh
npm run test
```

### Coverage

```sh
npm run coverage
```

## Demo

http://uxcore.github.io/components/list

## Contribute

Yes please! See the [CONTRIBUTING](https://github.com/uxcore/uxcore/blob/master/CONTRIBUTING.md) for details.

## API

## Props

| Name | Type | Required | Default | Since | Comments |
|---|---|---|---|---|---|
|width               |number              |optional  |auto        | -         |表格的宽度|
|height              |number              |optional  |auto        | -         |表格的高度|
|showPager           |boolean             |optional  |true        | -         |是否显示分页|
|showPagerTotal      |boolean             |optional  |false       | -         |是否显示分页的总数部分|
|showSearch          |boolean             |optional  |false       | -         |是否显示内置的搜索栏| 
|locale              |string              |optional  |zh-cn       | -         |国家化，目前支持 zh-cn/en-us|
|beforeFetch         |function(data, from)|optional  |noop        | -         |两个参数，data 表示表格请求数据时即将发送的参数，from 表示这次请求数据的行为从哪里产生，内置的有 `search`(搜索栏),`order`(排序) & `pagination`(分页)，该函数需要返回值，返回值为真正请求所携带的参数。|
|processData         |function(data)      |optional  |noop        | -         |有时源返回的数据格式，并不符合 Table 的要求，可以通过此函数进行调整，参数 data 是返回数据中 content 字段的 value，该函数需要返回值，返回值为符合 content 字段 value 的数据结构。|
|pageSize            |number              |optional  |10          | -         |每页显示多少条数据|
|pagerSizeOptions    |array               |optional  |[10,20,30,40] | -       |显示的可选 pageSize|
|data                |object              |optional  |-           | -         |在远端数据还没有返回时用作默认数据|
|fetchUrl            |string              |optional  |""          | -         |表格的数据源|
|fetchParams         |object              |optional  |-           | -         |表格在请求数据时，会额外附带的参数，具有最高的优先级|
|actionBar           |object/array        |optional  |null        | -         |表格内置的操作条配置，详细[见此](#actionbar)|
|onFetchError        |function(result)    |optional  |noop        | -         |在返回数据中 success 不是 true 的情况下触发，返回所有请求得到的数据|

## 返回的数据格式

* 数据格式的约定[见此](http://gitlab.alibaba-inc.com/alinw/yosemite/issues/18) 

```javascript
   {
    "content":{
        "data":[
            {   
                "id":'1'
                "grade":"grade1",
                "email":"email1",
                "firstName":"firstName1",
                "lastName":"lastName1",
                "birthDate":"birthDate1",
                "country":"country1",
                "city":"city1"
            }
            ...
    
        ],
        "currentPage":1,
        "totalCount":30
    },
    "success": true,
    "errorCode": "",
    "errorMsg": ""
    }

```

> 上面的数据格式是 ajax 返回的数据格式要求，如果你通过 jsxdata 传值，只需要 content 里面的内容。

```javascript
{
    "data":[
        {   
            "id":'1'
            "grade":"grade1",
            "email":"email1",
            "firstName":"firstName1",
            "lastName":"lastName1",
            "birthDate":"birthDate1",
            "country":"country1",
            "city":"city1"
        }
        ...

    ],
    "currentPage":1,
    "totalCount":30
}
```

### ActionBar 配置的例子

```javascript

// actionBar 支持传入一个对象
actionBar: {
    "新增行": () => { // 点击回调
        me.refs.grid.addEmptyRow();
    },
    "编辑所有行": () => {
        me.refs.grid.editAllRow();
    }
}

// 或者定制能力更加强大的数组
actionBar: [
    {
        title: '新增行', // 显示名称
        callback: () => { // 点击回调
            me.refs.grid.addEmptyRow();
        },
        render: (title) => { // 定制渲染
            return <Button>{title}</Button>
        }
    },
    {
        title: "编辑所有行",
        callback: () => {
            me.refs.grid.editAllRow();
        }
    },
    {
        title: "保存所有行",
        callback: () => {
            me.refs.grid.saveAllRow();
        }
    }
    
]
```

HISTORY.md
# history
## 0.1.8

* `CHANGED` support js style export
## 0.1.7

* `CHANGED` update to react 16.x
## 0.1.6

* `CHANGED` update to react 15.x
## 0.1.5

* `CHANGED` replace deepcopy with lodash/cloneDeep
* `FIXED` setState may be triggered in componentWillMount

## 0.1.4

* `FIXED` add currentPage default value

## 0.1.3

* `FIXED` fix processData bug

## 0.1.2

* `CHANGED` adjust style.


src/List.js
/**
 * List Component for uxcore
 * @author eternalsky
 *
 * Copyright 2015-2016, Uxcore Team, Alinw.
 * All rights reserved.
 */
import React from 'react';

import PropTypes from 'prop-types';
import ActionBar from './ActionBar';
import ListCore from './ListCore';
import Pagination from 'uxcore-pagination';
import deepcopy from 'lodash/cloneDeep';
import isEqual from 'lodash/isEqual';
import NattyFetch from 'natty-fetch';
import assign from 'object-assign';
import classnames from 'classnames';
import Promise from 'lie';

class List extends React.Component {
  constructor(props) {
    super(props);
    this.data = deepcopy(props.data);
    this.handleActionBarSearch = this.handleActionBarSearch.bind(this);
    this.handlePagerChange = this.handlePagerChange.bind(this);
    this.handlePagerSizeChange = this.handlePagerSizeChange.bind(this);
    this.processData = this.processData.bind(this);
    this.state = {
      data: this.data,
      pageSize: props.pageSize,
    };
  }

  componentDidMount() {
    this.fetchData();
  }

  componentWillReceiveProps(nextProps) {
    const newState = {};
    if (!isEqual(nextProps.data, this.data)) {
      this.data = deepcopy(nextProps.data);
      newState.data = this.data;
    }
    if (nextProps.pageSize !== this.props.pageSize) {
      newState.pageSize = nextProps.pageSize;
    }
    this.setState(newState);
  }

  componentWillUnmount() {
    const me = this;
    if (me.fetch) {
      me.fetch.abort();
    }
  }

  /**
   * only call this method in fetchData()
   */
  processData(data) {
    const me = this;
    const { processData } = me.props;
    return processData(data);
  }

  handleActionBarSearch(key) {
    const me = this;
    me.setState({
      searchTxt: key,
    }, () => {
      me.fetchData('search');
    });
  }

  handlePagerSizeChange(currentPage, pageSize) {
    const me = this;
    me.setState({
      pageSize,
    }, () => {
      me.fetchData('pagination');
    });
  }

  handlePagerChange(currentPage) {
    const me = this;
    const data = assign({}, me.state.data);
    data.currentPage = currentPage;
    me.setState({
      data,
    }, () => {
      me.fetchData('pagination');
    });
  }

  fetchData(from) {
    const me = this;
    const { fetchUrl, fitResponse, beforeFetch, onFetchError, fetchParams } = me.props;
    const { data, pageSize, searchTxt } = me.state;
    const isJsonp = me.props.isJsonp === undefined ? /\.jsonp/.test(fetchUrl) : me.props.isJsonp;
    const fetchData = assign({
      currentPage: data.currentPage || 1,
      pageSize,
      searchTxt,
    }, fetchParams);
    if (fetchUrl) {
      me.fetch = NattyFetch.create({
        url: fetchUrl,
        data: beforeFetch(fetchData, from),
        fit: response => fitResponse(response),
        process: me.processData,
        jsonp: isJsonp,
        Promise,
      });
      me.fetch().then((content) => {
        me.setState({
          data: content,
        });
      }).catch((err) => {
        onFetchError(err);
      });
    }
  }

  renderActionBar() {
    const me = this;
    const {
      showSearch,
      actionBar,
      searchBarPlaceholder,
    } = me.props;
    const actionBarProps = {
      onSearch: this.handleActionBarSearch,
      key: 'actionbar',
      actionBarConfig: actionBar,
      showSearch,
      searchBarPlaceholder,
    };
    return <ActionBar {...actionBarProps} />;
  }

  renderListCore() {
    const me = this;
    const data = me.state.data || {};
    return (
      <ListCore
        data={data.data}
        renderItem={me.props.renderItem}
        locale={me.props.locale}
      />
    );
  }

  renderPager() {
    const me = this;
    const { data } = me.state;
    const { prefixCls, showPager, locale, showPagerTotal, pageSize, pagerSizeOptions } = me.props;
    /**
     *
     * data structure
     * {
     *   data: [],
     *   currentPage: 1,
     *   totalCount: 30
     * }
     *
     */
    if (showPager && data && data.totalCount && data.currentPage) {
      return (
        <div className={`${prefixCls}-pager`}>
          <Pagination
            total={data.totalCount}
            locale={locale}
            showSizeChanger
            showQuickJumper
            showTotal={showPagerTotal}
            total={data.totalCount}
            onShowSizeChange={me.handlePagerSizeChange}
            onChange={me.handlePagerChange}
            current={data.currentPage}
            pageSize={pageSize}
            sizeOptions={pagerSizeOptions}
          />
        </div>
      );
    }
    return null;
  }

  render() {
    const me = this;
    const { prefixCls, width, height, className } = me.props;
    return (
      <div
        className={classnames({
          [`${prefixCls}`]: true,
          [className]: !!className,
        })}
        style={{ width, height }}
      >
        {me.renderActionBar()}
        {me.renderListCore()}
        {me.renderPager()}
      </div>
    );
  }
}

List.defaultProps = {
  prefixCls: 'kuma-list',
  locale: 'zh-cn',
  fetchParams: {},
  searchBarPlaceholder: '请输入',
  showPagerTotal: true,
  showSearch: false,
  showPager: true,
  pageSize: 10,
  pagerSizeOptions: [10, 20, 30, 40],
  data: {},
  renderItem: () => {},
  fitResponse: response => response,
  beforeFetch: data => data,
  processData: data => data,
  onFetchError: (err) => { console.error(err.stack); },
  width: 'auto',
  height: 'auto',
};


// http://facebook.github.io/react/docs/reusable-components.html
List.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  locale: PropTypes.string,
  emptyText: PropTypes.string,
  fetchUrl: PropTypes.string,
  fetchParams: PropTypes.object,
  searchBarPlaceholder: PropTypes.string,
  data: PropTypes.object,
  renderItem: PropTypes.func,
  fitResponse: PropTypes.func,
  beforeFetch: PropTypes.func,
  processData: PropTypes.func,
  onFetchError: PropTypes.func,
  isJsonp: PropTypes.bool,
  showSearch: PropTypes.bool,
  showPager: PropTypes.bool,
  showPagerTotal: PropTypes.bool,
  pagerSizeOptions: PropTypes.array,
  pageSize: PropTypes.number,
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  height: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  actionBar: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.object,
  ]),
};

List.displayName = 'List';

export default List;
